### DISCUSSION TITLE
How to improve this `ORDER BY <column> DESC LIMIT 1` query?

### DISCUSSION THREAD

**Discussion by paxw-panevo** [ClickHouse/ClickHouse#71262] [Q&A] [answered]
I have a table with `ORDER BY Keys`: c_id, v_id, datetime. When I run the query `SELECT datetime, val FROM table_name FINAL WHERE c_id = 1 AND v_id = 1 AND datetime < 1726625761 ORDER BY datetime DESC LIMIT 1` it reads millions of rows (all the data of `c_id=1;v_id=1` from the beginning of time up to 1726625761) and it's consuming ~200MB of memory.

Is there a way to prevent the row from reading millions especially since we're only interested in the "top" value?

Our table's `SHOW CREATE` statement is:

```
statement: CREATE TABLE default.variables
(
    `c_id` UInt32,
    `v_id` UInt32,
    `datetime` DateTime('UTC'),
    `val` Float64
)
ENGINE = SharedReplacingMergeTree('/clickhouse/tables/{uuid}/{shard}', '{replica}')
PRIMARY KEY (c_id, v_id, datetime)
ORDER BY (c_id, v_id, datetime)
SETTINGS index_granularity = 8192
```

**Comment by UnamedRus** [comment, marked as answer]
You don't need FINAL here, i think

Can you run

```
SELECT datetime, val FROM table_name WHERE c_id = 1 AND v_id = 1 AND datetime < 1726625761 ORDER BY datetime DESC LIMIT 1
```

Or

```
SELECT datetime, val FROM table_name WHERE c_id = 1 AND v_id = 1 AND datetime < 1726625761 ORDER BY c_id  DESC, v_id  DESC, datetime DESC LIMIT 1
```

And check that it actually use read_in_order ?

(
You can enable logs via

```
set send_logs_level='trace';
```

**Reply by paxw-panevo** [reply]
Thanks so much @UnamedRus! 

Removing `FINAL` does indeed help and I can see that it now only reads 40K rows instead of 9M. We did not explicitly disable `optimize_read_in_order` anywhere so I assume removing `FINAL` allows ClickHouse to perform the optimize_read_in_order optimization automatically. I can't tell from the logs (I'm new to reading ClickHouse trace logs) whether ClickHouse was using read_in_order but I read [from the docs](https://clickhouse.com/docs/en/sql-reference/statements/select/order-by#optimization-of-data-reading) that this optimization will only work if the query does not use FINAL.

Some things I've observed using the trace logs,

With filesystem caching disabled:
   - the processing time between the 2 queries are comparable, without FINAL being slightly faster. 0.642956295 sec. (without FINAL) and  0.771587869 sec. (with FINAL)
   - the number of rows read and peak memory usage of the query without FINAL is significantly lower: Read 40_960 rows, Peak memory usage (for query): 2.78 MiB. Compare this to the query with FINAL:  Read 9_699_328 rows, Peak memory usage (for query): 30.55 MiB.

With filesystem caching enabled:
  - the processing time without FINAL is significantly faster: 0.007333733 sec. vs 0.34286963 sec.
  - peak memory use without FINAL is lower: 19.64 MiB vs 42.55 MiB

Some questions I have:

1. Is this understanding correct: I don't need to use the FINAL keyword for this specific query with a `LIMIT 1` clause. If there are 2 (or 3, etc.) values for timestamp X, ClickHouse will naturally return the last row inserted. The FINAL keyword is relevant if the query involves returning multiple rows since the FINAL keyword (along with the underlying ReplacingMergeTree engine) will do a merge during query runtime, removing duplicate rows based on ORDER BY keys.

2. Looking at the logs for the query without the FINAL keyword, I noticed these operations that are not found in the logs of the query with the FINAL keyword:
- QueryPlanOptimizePrewhere
- MergeTreeSelectProcessor

I also noticed that without the FINAL keyword, ranges are read _in reverse order_ (I assume this means in DESC order, as specified in my query): `Reading 1 ranges in reverse order from part...`. Whereas with the FINAL keyword, ranges are read _in order_: `Reading 1 ranges in order from part ...`

<details>
<summary>some of the logs</summary>

```
...
<Trace> Planner: Query to stage Complete
<Trace> Planner: Query from stage FetchColumns to stage Complete
<Trace> QueryPlanOptimizePrewhere: The min valid primary key position for moving to the tail of PREWHERE is 2
<Trace> QueryPlanOptimizePrewhere: Condition less(datetime, 1726625761_UInt32) moved to PREWHERE
<Trace> QueryPlanOptimizePrewhere: Condition equals(v_id, 3112_UInt16) moved to PREWHERE
<Trace> QueryPlanOptimizePrewhere: Condition equals(c_id, 4_UInt8) moved to PREWHERE
<Debug> default.variables (28ec7347-12e6-4782-a319-f798f8327275) (SelectExecutor): Key condition: (column 0 in [4, 4]), (column 1 in [3112, 3112]), and, (column 2 in (-Inf, 1726625760]), and
<Trace> default.variables (28ec7347-12e6-4782-a319-f798f8327275) (SelectExecutor): Filtering marks by primary and secondary keys
...
<Trace> default.variables (28ec7347-12e6-4782-a319-f798f8327275) (SelectExecutor): Reading 1 ranges in reverse order from part all_635108_703658_690_689681, approx. 8192 rows starting from 8192
<Trace> MergeTreeSelectProcessor: PREWHERE condition was split into 3 steps: "equals(__table1.c_id, 4_UInt8)", "equals(__table1.v_id, 3112_UInt16)", "less(__table1.datetime, 1726625761_UInt32)"
<Trace> default.variables (28ec7347-12e6-4782-a319-f798f8327275) (SelectExecutor): Reading 16 ranges in reverse order from part all_168297_635107_4312_689681, approx. 876544 rows starting from 8749056
<Trace> MergeTreeSelectProcessor: PREWHERE condition was split into 3 steps: "equals(__table1.c_id, 4_UInt8)", "equals(__table1.v_id, 3112_UInt16)", "less(__table1.datetime, 1726625761_UInt32)"
<Trace> default.variables (28ec7347-12e6-4782-a319-f798f8327275) (SelectExecutor): Reading 14 ranges in reverse order from part all_102842_168296_1201_689681, approx. 745472 rows starting from 1359872
<Trace> MergeTreeSelectProcessor: PREWHERE condition was split into 3 steps: "equals(__table1.c_id, 4_UInt8)", "equals(__table1.v_id, 3112_UInt16)", "less(__table1.datetime, 1726625761_UInt32)"
<Trace> default.variables (28ec7347-12e6-4782-a319-f798f8327275) (SelectExecutor): Reading 51 ranges in reverse order from part all_0_102841_490_689681, approx. 3203072 rows starting from 493772800
<Trace> MergeTreeSelectProcessor: PREWHERE condition was split into 3 steps: "equals(__table1.c_id, 4_UInt8)", "equals(__table1.v_id, 3112_UInt16)", "less(__table1.datetime, 1726625761_UInt32)"
<Trace> default.variables (28ec7347-12e6-4782-a319-f798f8327275) (SelectExecutor): Reading 76 ranges in reverse order from part all_0_102841_490_689681, approx. 4833280 rows starting from 496975872
<Trace> MergeTreeSelectProcessor: PREWHERE condition was split into 3 steps: "equals(__table1.c_id, 4_UInt8)", "equals(__table1.v_id, 3112_UInt16)", "less(__table1.datetime, 1726625761_UInt32)"
<Debug> MergingSortedTransform: Merge sorted 1 blocks, 7979 rows in 0.756008688 sec., 10554.111515713164 rows/sec., 123.68 KiB/sec
...
```
</details>

Are these hints that ClickHouse is doing a read_in_order optimization? I only assume it is based on [what the docs say about the optimization](https://clickhouse.com/docs/en/sql-reference/statements/select/order-by#optimization-of-data-reading) and their illustration as well of [how the optimization works](https://clickhouse.com/docs/knowledgebase/async_vs_optimize_read_in_order#what-about-optimize_read_in_order).

3. Do you have any recommended resource / specific class from the ClickHouse catalog that would help me read clickhouse trace logs?

Thanks again @UnamedRus! I appreciate it so much!
